### PR TITLE
DXCDT-388: Enabling preservation of array replace markers in YAML

### PR DESCRIPTION
**Pull request by willvedd**
### 🔧 Changes

One of the final pieces for delivery a MVP deliverable for keyword preservation on export is enabling the preservation of array replace keywords (ex: `@@LOGOUT_URLS@@`) in YAML configuration formats. The hurdle with this limitation can be broken down into two major problems:

**Problem Statement 1:**
When using array replace syntax in a YAML configuration file, the markers can be expressed as-is without quotes, which is invalid YAML. There is a built-in YAML to JSON conversion step that will fail. Example of an error-causing configuration:

```yaml
tenant:
  allowed_logout_urls: @@LOGOUT_URLS@@
```

**Solution:**
Introduce an intermediate step before the YAML parsing that looks for unquoted array replace markers and upon finding them, wrap in quotes. Array replace markers that are already wrapped in quotes should be left as-is. 

--- 

**Problem Statement 2:**
When using array replace syntax in a YAML configuration file, the markers can be expressed as-is without quotes, wrapped with double quotes or wrapped in single quotes. While technically always true, the keyword replacement process preserves these array values wrapped in single-quotes.

For example, all of the following are valid configuration files:
```yaml
tenant:
  allowed_logout_urls: @@LOGOUT_URLS@@
```
```yaml
tenant:
  allowed_logout_urls: '@@LOGOUT_URLS@@'
```
```yaml
tenant:
  allowed_logout_urls: "@@LOGOUT_URLS@@"
```

**Solution:**

Appropriately handle each scenario during the keyword replacement step. In practice, this means introducing a third regex replacement step for single-quoted strings in addition to the double-quoted and unquoted array replace markers.

---

Together, the combination of these solutions enabled a seamless experience for users needing to preserve array replace keywords in the YAML format.

### 📚 References

Related RFC: #688 

### 🔬 Testing

Added new unit tests for any functions that were altered. Additionally, added a couple assertions to E2E tests. Finally, did quite a bit of manual testing for another layer of verification.

### 📝 Checklist

- [x] All new/changed/fixed functionality is covered by tests (or N/A)
- [x] I have added documentation for all new/changed functionality (or N/A)

### CODE REVIEW

File: src/context/yaml/handlers/actions.ts [hunk 1]
```diff
@@ -86,7 +86,8 @@ async function dump(context: YAMLContext): Promise<ParsedActions> {
       runtime: action.runtime,
       dependencies: action.dependencies || [],
       status: action.status,
-      secrets: mapSecrets(action.secrets || []),
+      secrets:
+        typeof action.secrets === 'string' ? action.secrets : mapSecrets(action.secrets || []), //Enables keyword preservation to operate on action secrets
```
Comment (willvedd) [inline src/context/yaml/handlers/actions.ts:90]: Most array preservation will work as-is for all YAML handlers, but there is this one instance where a string value for `action.secrets` would trigger a runtime error when applying a map function.

File: src/tools/utils.ts [hunk 2]
```diff
@@ -8,9 +8,11 @@ import constants from './constants';
 
 export const keywordReplaceArrayRegExp = (key) => {
   const pattern = `@@${key}@@`;
-  const patternWithQuotes = `"${pattern}"`;
+  //YAML format supports both single and double quotes for strings
+  const patternWithSingleQuotes = `'${pattern}'`;
+  const patternWithDoubleQuotes = `"${pattern}"`;
 
-  return new RegExp(`${patternWithQuotes}|${pattern}`, 'g');
+  return new RegExp(`${patternWithSingleQuotes}|${patternWithDoubleQuotes}|${pattern}`, 'g');
```
Comment (willvedd) [inline src/tools/utils.ts:15]: Array replace markers could be wrapped in single quotes, double quotes or not wrapped at all. When performing this replacement, it is necessary to check for all three.

File: src/tools/utils.ts [hunk 3]
```diff
@@ -45,6 +47,20 @@ export function keywordReplace(input: string, mappings: KeywordMappings): string
   return input;
 }
 
+// wrapArrayReplaceMarkersInStrings will wrap array replacement markers in quotes.
+// This is necessary for YAML format in the context of keyword replacement
+// to preserve the keyword markers while also maintaining valid YAML syntax.
+export function wrapArrayReplaceMarkersInStrings(body: string, mappings: KeywordMappings): string {
```
Comment (willvedd) [inline src/tools/utils.ts:53]: This function does the heavy lifting in this PR. It solves the problem outlined in Problem 1 of the description.

File: test/context/yaml/context.test.js [hunk 4]
```diff
@@ -541,23 +541,18 @@ describe('#YAML context validation', () => {
...
+      friendly_name: "##ENV## Tenant"
+      enabled_locales: @@LANGUAGES@@
+    connections:
+      - name: connection-1
+        strategy: waad
+        options:
+          tenant_domain: "##DOMAIN##"
+    `;
+
+    fs.writeFileSync(tenantFile, localAssets);
```
Comment (willvedd) [inline test/context/yaml/context.test.js:555]: Not possible to test this scenario with JSON, needs to be converted to YAML. Note the unwrapped `@@LANGUAGES@@` array replace instance.

**Review by sergiught — APPROVED** [review]

**Pull request merged**